File: leetcode/0328-odd-even-linked-list/solution.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    /**
     * RUNTIME: 12 ms (88.18%)
     * MEMORY: 11.1 MB (5.19%)
    */
    ListNode* oddEvenList(ListNode* head) {
        if (!head) return NULL;
        if (head->next == NULL || head->next->next == NULL) return head;
        ListNode *result = new ListNode();
        ListNode *res_copy = result, *current = head;
        while (true) {
            res_copy->val = current->val;
            ListNode *tmp = new ListNode();
            res_copy->next = tmp;
            if (current->next == NULL || current->next->next == NULL) { break; }
            current = current->next->next;
            res_copy = res_copy->next;
        }
        current = head->next;
        while (true) {
            ListNode *tmp = new ListNode(current->val);
            res_copy->next = tmp;
            if (current->next == NULL || current->next->next == NULL) { break; }
            current = current->next->next;
            res_copy = res_copy->next;
        }
        return result;
    }
};

File: leetcode/0328-odd-even-linked-list/solution.py
# Definition for singly-linked list.
class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next

class Solution:
    def oddEvenList(self, head: Optional[ListNode]) -> Optional[ListNode]:
        """
        RUNTIME: 46 ms (88.93%)
        MEMORY: 17.1 MB (7.51%)
        """
        if not head: return None
        if not head.next or not head.next.next: return head
        result = ListNode()
        res_copy = result
        current = head
        while True:
            res_copy.val = current.val
            res_copy.next = ListNode()
            if not current.next or not current.next.next: break
            current = current.next.next
            res_copy = res_copy.next
        current = head.next
        while True:
            res_copy.next = ListNode(current.val)
            if not current.next or not current.next.next: break
            current = current.next.next
            res_copy = res_copy.next
        return result
